src/linter/block.go

```diff
@@ -723,16 +726,112 @@ func (b *blockWalker) handleTry(s *ir.TryStmt) bool {
 		})
 	}
 
-	if othersExit && ctx.exitFlags != 0 {
+	if othersExit && tryCtx.exitFlags != 0 {
 		b.ctx.exitFlags |= prematureExitFlags
-		b.ctx.exitFlags |= ctx.exitFlags
+		b.ctx.exitFlags |= tryCtx.exitFlags
 	}
 
-	b.ctx.containsExitFlags |= ctx.containsExitFlags
+	b.ctx.containsExitFlags |= tryCtx.containsExitFlags
 
 	return false
 }
 
+func (b *blockWalker) checkUnreachableForFinallyReturn(tryStmt *ir.TryStmt, tryCtx *blockContext, finallyCtx *blockContext, catchContexts []*blockContext) {
+	if finallyCtx == nil {
+		return
+	}
+
+	var exitPoints []exitPoint
+
+	// if try contains some other return/die/etc statement
+	if tryCtx.containsExitFlags != 0 && tryCtx.containsExitFlags^FlagThrow != 0 {
+		points := b.findExitPointsByFlags(&ir.StmtList{Stmts: tryStmt.Stmts}, tryCtx.containsExitFlags^FlagThrow)
+		exitPoints = append(exitPoints, points...)
+	}
+
+	var catchContainsDie bool
+	var catchWithDieIndex int
+
+	for i, context := range catchContexts {
+		if context.exitFlags != 0 || context.containsExitFlags != 0 {
+			points := b.findExitPointsByFlags(tryStmt.Catches[i], context.exitFlags^FlagDie)
+			exitPoints = append(exitPoints, points...)
+		}
+
+		if context.exitFlags&FlagDie == FlagDie {
+			catchContainsDie = true
+			catchWithDieIndex = i + 1
+		}
+	}
+
+	if catchContainsDie {
+		b.r.Report(tryStmt.Finally, LevelError, "unreachable", "block finally is unreachable (because %d catch block contains a exit/die)", catchWithDieIndex)
+
+		// If there is an error when the finally block is unreachable,
+		// then errors due to return in finally are skipped.
+		return
+	}
+
+	if finallyCtx.exitFlags == FlagReturn {
+		var finallyReturnPos position.Position
+		finallyReturns := b.findExitPointsByFlags(tryStmt.Finally, FlagReturn)
+		if len(finallyReturns) > 0 {
+			finallyReturnPos = *ir.GetPosition(finallyReturns[0].n)
+		}
+
+		for _, point := range exitPoints {
+			b.r.Report(point.n, LevelError, "unreachable", "%s is unreachable (because finally block contains a return on line %d)", point.typ, finallyReturnPos.StartLine)
+		}
+	}
+}
+
+type exitPoint struct {
+	n   ir.Node
+	typ string
+}
+
+func (b *blockWalker) findExitPointsByFlags(where ir.Node, exitFlags int) (points []exitPoint) {
+	irutil.Inspect(where, func(n ir.Node) bool {
+		if exitFlags&FlagReturn == FlagReturn {
```

Comment:
Do you want to check that this bit is set?
Why not do it like this:
```go
exitFlags&FlagReturn != 0
```
It's less error-prone after the copy/paste.

Reply:
Fixed.